# Datentypen für Dezimalzahlen

<div class="prereq">
    <h3>Was man wissen sollte</h3>
    <div>
        <a class="prereq" href="/user-redirect/algoviz/lessons/02_Grundlagen/01_Datentypen.ipynb">Datentypen</a>
    </div>
</div>

<div class="slideshow /02_Grundlagen/07_Floats/slides.json">Datentypen für Dezimalzahlen</div>

In [1]:
#include <iostream>

## Ein paar Experimente mit floats und doubles

Probieren wir als erstes die kleinsten positiven Werte aus.

In [2]:
float minFloat = 1.17549e-038;
double minDouble = 4.940656458412e-324;

std::cout << minFloat << "\n";
std::cout << minDouble << "\n";

1.17549e-38
4.94066e-324


Weisen wir jetzt den kleinste `double` Wert einem `float`zu:

In [3]:
float klein = minDouble;

std::cout << klein << std::endl;

0


Wie man sieht wird der Wert auf 0 gerundet.

In [4]:
long lang = 16777219;
float gross = lang;

std::cout << lang << std::endl;
std::cout << gross << std::endl;

16777219
1.67772e+07


Wie es scheint haben wir in `gross` jetzt den Wert 16777200 statt 1677219. Aber es gibt noch ein Problem: Es werden nur fünf Stellen nach dem Komma angezeigt. Das kann man mit `std::setprecision(digits)` aber verändern. Fügt man es in die Ausgabe ein, werden die folgenden Gleitkommatypen mit der angegebenen Anzahl von Nachkommastellen ausgegeben.

In [5]:
std::cout << std::setprecision(10) << gross << std::endl;

16777220


Da ist die 16777220, wie in der Slideshow angegeben. Wie man sieht, wird durch die höher gestellte Zahl auch die sogenannte **wissenschaftliche Notation** mit Exponenten aufgelöst.

Spielen wir noch etwas damit herum:

In [9]:
float zahl = 1234567890.1234567890;

std::cout << std::setprecision(0) << zahl << std::endl;
std::cout << std::setprecision(1) << zahl << std::endl;
std::cout << std::setprecision(3) << zahl << std::endl;
std::cout << std::setprecision(5) << zahl << std::endl;
std::cout << std::setprecision(7) << zahl << std::endl;
std::cout << std::setprecision(9) << zahl << std::endl;
std::cout << std::setprecision(10) << zahl << std::endl;

1e+09
1e+09
1.23e+09
1.2346e+09
1.234568e+09
1.23456794e+09
1234567936


Wie man sieht, wurden die Nachkommastellen "vergessen" und die letzten drei Ziffern stimmen nicht mehr. 
Das liegt einfach daran, dass der Wert 1234567890 bereits zu groß für eine Präzision unter 1 ist.

Machen wir das Ganze mit einem `double` sieht es so aus:

In [10]:
double praezise = 1234567890.1234567890;

std::cout << std::setprecision(0) << praezise << std::endl;
std::cout << std::setprecision(1) << praezise << std::endl;
std::cout << std::setprecision(3) << praezise << std::endl;
std::cout << std::setprecision(5) << praezise << std::endl;
std::cout << std::setprecision(7) << praezise << std::endl;
std::cout << std::setprecision(9) << praezise << std::endl;
std::cout << std::setprecision(10) << praezise << std::endl;

1e+09
1e+09
1.23e+09
1.2346e+09
1.234568e+09
1.23456789e+09
1234567890


Jetzt stimmt das Resultat.

<div class="followup">
    <h3>Weitere Informationen</h3>
    <div>
        Wer mehr über die Speicherung von Dezimalzahlen erfahren möchte, sollte mit <a class="addon" href="/user-redirect/algoviz/lessons/02_Grundlagen/08_Festkomma.ipynb">Festkommazahlen</a> fortfahren.
    </div>
</div>